### PR TITLE
provider/aws: Added Lambda Environment configuration (supersedes #10255)

### PR DESCRIPTION
**Pull request by catsby**
Continues #10255 with some touchups, namely `environment` wasn't being read. I also encountered an issue where the API would return an "empty" vpc config, so now we filter those out instead of populating the state with an empty config, triggering a `ForceNew`.

@Ninir please take a look!

```
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSLambdaFunction_en -timeout 120m
=== RUN   TestAccAWSLambdaFunction_envVariables
--- PASS: TestAccAWSLambdaFunction_envVariables (33.15s)
=== RUN   TestAccAWSLambdaFunction_encryptedEnvVariables
--- PASS: TestAccAWSLambdaFunction_encryptedEnvVariables (72.76s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	105.929s
```

### CODE REVIEW

File: builtin/providers/aws/resource_aws_lambda_function.go [hunk 1]
```diff
@@ -323,15 +323,26 @@ func resourceAwsLambdaFunctionRead(d *schema.ResourceData, meta interface{}) err
...
-		err := d.Set("vpc_config", config)
-		if err != nil {
-			return fmt.Errorf("Failed setting vpc_config: %s", err)
-		}
+
+	config := flattenLambdaVpcConfigResponse(function.VpcConfig)
+	log.Printf("[INFO] Setting Lambda %s VPC config %#v from API", d.Id(), config)
+	vpcSetErr := d.Set("vpc_config", config)
+	if vpcSetErr != nil {
+		return fmt.Errorf("Failed setting vpc_config: %s", vpcSetErr)
```
Comment (Ninir) [inline builtin/providers/aws/resource_aws_lambda_function.go]: Could you share the config you used to find that please? thanks!

Comment (Ninir) [inline builtin/providers/aws/resource_aws_lambda_function.go]: Oh got it! good catch :)
(I would have use my own PR to reproduce, but still, thanks!)

File: builtin/providers/aws/resource_aws_lambda_function.go [hunk 2]
```diff
@@ -282,15 +322,27 @@ func resourceAwsLambdaFunctionRead(d *schema.ResourceData, meta interface{}) err
...
 	}
+
 	d.Set("source_code_hash", function.CodeSha256)
 
+	var envSetErr error
+	if function.Environment == nil {
+		envSetErr = d.Set("environment", []interface{}{})
+	} else {
+		envSetErr = d.Set("environment", flattenLambdaEnvironment(function.Environment.Variables))
+	}
```
Comment (Ninir) [inline builtin/providers/aws/resource_aws_lambda_function.go]: I think I forgot this part... Newbie mistake... 👍

**Comment by Ninir** [comment]
Let's go @catsby ! thanks for the help and review on this, learned a lot as a newcomer :)

File: builtin/providers/aws/resource_aws_lambda_function.go (hunk 1 above)
Comment (catsby) [inline builtin/providers/aws/resource_aws_lambda_function.go]: Example DEBUG output of said empty vpc:

```
2016/11/21 17:03:41 [DEBUG] plugin: terraform-provider-aws: Function:
2016/11/21 17:03:41 [DEBUG] plugin: terraform-provider-aws: {
2016/11/21 17:03:41 [DEBUG] plugin: terraform-provider-aws:   CodeSha256: "Ux/n9CP8l+7Ht0tICw0QPs0yLdC1b+1nJ9K5MZR9ENw=",
					[ ... ]
2016/11/21 17:03:41 [DEBUG] plugin: terraform-provider-aws:   Role: "arn:aws:iam::187416307283:role/cts-iam-lambda",
2016/11/21 17:03:41 [DEBUG] plugin: terraform-provider-aws:   Runtime: "nodejs",
2016/11/21 17:03:41 [DEBUG] plugin: terraform-provider-aws:   Timeout: 3,
2016/11/21 17:03:41 [DEBUG] plugin: terraform-provider-aws:   Version: "$LATEST",
2016/11/21 17:03:41 [DEBUG] plugin: terraform-provider-aws:   VpcConfig: {
2016/11/21 17:03:41 [DEBUG] plugin: terraform-provider-aws:     SecurityGroupIds: [],
2016/11/21 17:03:41 [DEBUG] plugin: terraform-provider-aws:     SubnetIds: [],
2016/11/21 17:03:41 [DEBUG] plugin: terraform-provider-aws:     VpcId: ""
2016/11/21 17:03:41 [DEBUG] plugin: terraform-provider-aws:   }
2016/11/21 17:03:41 [DEBUG] plugin: terraform-provider-aws: }
```

Only happens if I edit the ENV vars in the Web ui (in this case I added one)

Comment (catsby) [inline builtin/providers/aws/resource_aws_lambda_function.go]: > (I would have use my own PR to reproduce, but still, thanks!)

You still can! The only way I could reproduce is if I had 2+ env vars specified in config, and then used the web UI to remove one of them. Then my next plan would show the ForceNew diff as well as the config diff

**Comment by catsby** [comment]
Tests after a small refactor: 

```
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSLambdaFunction_en -timeout 120m
=== RUN   TestAccAWSLambdaFunction_envVariables
--- PASS: TestAccAWSLambdaFunction_envVariables (28.22s)
=== RUN   TestAccAWSLambdaFunction_encryptedEnvVariables
--- PASS: TestAccAWSLambdaFunction_encryptedEnvVariables (61.75s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	89.986s
Test:
``` 

merging, thanks @Ninir !

**Comment by kwilczynski** [comment]
@catsby @Ninir - are you going to update this PR with changes discussed from the other one? I don't want things from there to get lost :) Apologies!

**Comment by catsby** [comment]
@kwilczynski I'm not aware of them, and this PR has already been merged :/ can you open a new issue? Or new PR?

**Pull request merged**